src/filters/timestamp.rs

```diff
@@ -0,0 +1,234 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::sync::Arc;
+
+use chrono::prelude::*;
+use once_cell::sync::Lazy;
+use prometheus::HistogramVec;
+
+use crate::{
+    filters::prelude::*,
+    metadata::Value,
+    metrics::{
+        histogram_opts, registry, BUCKET_COUNT, BUCKET_FACTOR, BUCKET_START, DIRECTION_LABEL,
+        METADATA_KEY_LABEL, READ_DIRECTION_LABEL, WRITE_DIRECTION_LABEL,
+    },
+};
+
+crate::include_proto!("quilkin.filters.timestamp.v1alpha1");
+use self::quilkin::filters::timestamp::v1alpha1 as proto;
+
+pub(crate) static METRIC: Lazy<HistogramVec> = Lazy::new(|| {
+    prometheus::register_histogram_vec_with_registry! {
+        histogram_opts(
+            "seconds",
+            SUBSYSTEM,
+            "The duration of seconds of the `metadata_key` metric",
+            prometheus::exponential_buckets(BUCKET_START, BUCKET_FACTOR, BUCKET_COUNT).unwrap(),
+        ),
+        &[METADATA_KEY_LABEL, DIRECTION_LABEL],
+        registry(),
+    }
+    .unwrap()
+});
+
+const SUBSYSTEM: &str = "filters_timestamp";
+
+/// A filter that reads a metadata value as a timestamp to be observed in
+/// a histogram.
+#[derive(Debug, Clone)]
+pub struct Timestamp {
+    config: Arc<Config>,
+}
+
+impl Timestamp {
+    /// Observes the duration since a timestamp stored in `metadata` and now,
+    /// if present.
+    pub fn observe(
+        &self,
+        metadata: &crate::metadata::DynamicMetadata,
+        direction_label: &'static str,
+    ) {
+        let value = metadata
+            .get(&self.config.metadata_key)
+            .and_then(|item| match item {
+                Value::Number(item) => Some(*item as i64),
+                Value::Bytes(vec) => Some(i64::from_be_bytes((**vec).try_into().ok()?)),
+                _ => None,
+            });
+
+        let value = match value {
+            Some(item) => item,
+            None => return,
+        };
+
+        let naive = match NaiveDateTime::from_timestamp_opt(value, 0) {
+            Some(datetime) => datetime,
+            None => {
+                tracing::warn!(
+                    timestamp = value,
+                    metadata_key = &*self.config.metadata_key,
+                    "invalid unix timestamp"
+                );
+                return;
+            }
+        };
+
+        // Create a normal DateTime from the NaiveDateTime
+        let datetime: DateTime<Utc> = DateTime::from_utc(naive, Utc);
+
+        let now = Utc::now();
+        let seconds = now.signed_duration_since(datetime).num_seconds();
+        self.metric(direction_label).observe(seconds as f64);
```

Comment:
(non blocking question): Does it make sense to combine this with maxmind information somehow? Either in the label or in tracing information?

Just thinking about tracking ping round trip times using this filter, and wanting some granularity on where geographically people are at?

Reply:
> Does it make sense to combine this with maxmind information somehow? Either in the label or in tracing information?

Yes it does, I'm going to try to find a way to do purely through grafana before trying to integrate it directly in code.

Reply:
Make sense! 👍🏻